api/server.js

```diff
@@ -508,23 +512,83 @@ app.postJson('/api/v1/users', function(req, res) {
 });
 
 app.postJson('/api/v1/users/:username', function(req, res) {
-  auth.check(req, 'can_update_users', null, function(err) {
-    if (err) {
+  const username = req.params.username;
+
+  const updateUser = fullAccess =>
+    users.updateUser(username, req.body, fullAccess, (err, body) => {
+      if (err) {
+        return serverUtils.error(err, req, res, true);
+      }
+
+      res.json(body);
+    });
+
+  if (_.isEmpty(req.body)) {
+    return emptyJSONBodyError(req, res);
+  }
+
+  auth.check(req, 'can_update_users', null, err => {
+    if (err && err.statusCode === 503) {
       return serverUtils.error(err, req, res);
     }
-    if (_.isEmpty(req.body)) {
-      return emptyJSONBodyError(req, res);
+
+    if (!err) {
+      // Full access
+      return updateUser(true);
     }
-    users.updateUser(req.params.username, req.body, function(err, body) {
-      if (err) {
-        return serverUtils.error(err, req, res);
+
+    // Not full access, but maybe we're updating ourselves
+    auth.getUserCtx(req, (err, userCtx) => {
+      if (userCtx.name !== username) {
+        // Not updating ourselves, and we don't have permission to update others
+        return serverUtils.error({
+          message: 'You do not have permissions to modify this person',
+          code: 401
+        }, req, res, true);
+      }
+
+      // We're updating ourselves, but if we are updating our password we need
+      // to provide Basic Auth credentials to indicate we're entering the password
+      // fresh as of this connection
+      if (Object.keys(req.body).includes('password')) {
+        let credentials;
+        try {
+          credentials = auth.basicAuthCredentials(req);
+        } catch (err) {
+          err.statusCode = 401;
+          return serverUtils.error(err, req, res, true);
+        }
+
+        if (username !== credentials.username) {
+          // Make sure the Basic Auth credentials are actually for the same person
+          return serverUtils.error({
+            message: 'You do not have permissions to modify this person',
+            code: 401
+          }, req, res, true);
+        }
+
+        auth.validateBasicAuth(credentials, err => {
+          if (err) {
+            console.error(`Invalid authorization attempt on /api/v1/users/${username}`);
+            console.error(err);
+            return serverUtils.error({
+              message: 'To modifiy yourself you must provide valid Basic Auth credentials',
```

Comment:
I don't think this is leaking security information.

This is my understanding of the code:

If you've got this far then:
 - you are already authenticated, either with Basic Auth _or_ with a cookie (I don't actually know where this is happening, but I can't even get to line 515 without valid credentials)
 - the username you authenticated with is the same one you're trying to modify (you will get kicked at 564 otherwise)

You're already authenticated, this is just confirming to make sure the authentication was done with Basic Auth (ie if you provide a cookie and basic, maybe the basic is wrong the cookie was used instead).